Scale dry air mass on initalization for non-analytic, non-restart simulations

MiCurry opened the pull request:
This merge adds code to the MPAS dycore which allows the ability to scale the dry air mass for non-analytic and non-restart simulations. To enable dry air mass scaling, the namelist option, mpas_scale_dry_air_mass, which this merge also introduces, must be set to .true. in user_nl_cam.

This merge also adds a routine that can compute the global sum of a one dimensional real array across all tasks. For now, this single real routine is needed, but a future commit can add an additional routine to compute the global sum of a integer array and could add an abstract interface for both functions.

MiCurry committed:
Add a function to compute global sums of a real array

This commit adds a functions to the MPAS dycore subdriver that computes the
global sum of a real array across all tasks using the MPAS communicator
module, mpas_dmpar.

skamaroc committed:
Added the capability to scale the dry-air mass in the initialization.

MiCurry committed:
Update mpas global sum real subroutine

This commit updates the global sum real subroutine, cam_mpas_global_sum_real in
two ways. The first, is to be a function, rather than a subroutine with an out
argument and the second is that it removes the dminfo argument because there is
already a reference to dminfo within the cam_mpas_subdriver.

It also updates all of the calls to cam_mpas_global_sum_real.

In the furture, another commit could update cam_mpas_global_sum_real further by
adding a similar function to compute the global sum of a integer, and then add
a generic interface for the two.

MiCurry committed:
Add mpas_scale_dry_air_mass namelist option

This commit adds a new namelist option to CAM, mpas_scale_dry_air_mass.
Currently it is not used in the MPAS dycore, but it will be used to turn on dry
mass scaling. The current default is set to .false..

MiCurry committed:
Use mpas_scale_dry_air_mass namelist to turn on dry air mass scaling

This commit using the mpas_scale_dry_air_mass namelist option, which was added
in the last commit, to turn on the dry-air mass scaling.

MiCurry commented:
This PR is the same PR as #18, execpt it is now up-to-date with `cam_mpas_dev` and points to my new CAM fork, which was deleted so the upstream could be changed to ESCOMP.

jtruesdal commented on src/dynamics/mpas/dyn_comp.F90:
```diff
@@ -1028,6 +1045,69 @@ subroutine read_inidat(dyn_in)
+   !   test_value = sqrt(surface_integral/(4.0_r8*pi))
+   !   write(iulog,*) subname//': earth radius from area = ', test_value
+   ! end if
+
```
Remove dead code

MiCurry replied:
removed

jtruesdal commented on src/dynamics/mpas/dyn_comp.F90:
```diff
@@ -1028,6 +1045,69 @@ subroutine read_inidat(dyn_in)
+   if (mpas_scale_dry_air_mass) then
+
+      allocate( p_top(nCellsSolve), preliminary_dry_surface_pressure(nCellsSolve), pm(plev) )
+      ! (1) calculate pressure at the lid                                                                                                         
```
I would take this whole section and place it in a new routine.  SE and FV3 named similar routines set_dry_mass.  In addition, it needs to have log output similar to what SE and FV3 provide (ie. initial dry mass, scaled dry mass )

MiCurry replied:
Sounds good. I'll just create a new function in dyn_comp.

jtruesdal replied:
Great. Feel free to use the FV3 dyn_comp.F90 as a template for types of logging messages.  They will help us validate the additional functionality.  Sounds like Peter is OK with adding these pressures in physconst.F90, maybe ps_dry_topo, ps_dry_notopo ! CESM reference dry mass pressure with/without topography.  I can change them during the PR to cam_development if needed.

MiCurry replied:
Moved to a new function. And added ps_dry_topo/notopo to physconst.F90.

jtruesdal commented on src/dynamics/mpas/dyn_comp.F90:
```diff
@@ -1028,6 +1045,69 @@ subroutine read_inidat(dyn_in)
+      ! (5) reset qv to conserve mass                                                                                                             
+      tracers(ixqv,:,1:nCellsSolve) = tracers(ixqv,:,1:nCellsSolve)/scaling_ratio
+
+      ! (6) integrate down the column to compute full pressure given the density and qv                                                           
```
@PeterHjortLauritzen Do we have to worry about dry/wet tracers in this dycore.  ie conserve mass of wet tracers vs conserve mixing ration of dry?

PeterHjortLauritzen replied:
All mixing ratios in MPAS (incl. those on the initial condition file) are dry so no.

jtruesdal commented on src/dynamics/mpas/dyn_comp.F90:
```diff
@@ -1028,6 +1045,69 @@ subroutine read_inidat(dyn_in)
+      ! scaling_ratio = target_avg_dry_surface_pressure/preliminary_avg_dry_surface_pressure
+      scaling_ratio = 1.0
+      rho(:,:) = rho(:,:)*scaling_ratio
+
```
@PeterHjortLauritzen how about adding a dry mass reference pressure (98288.0) to physconst.F90 that all the dycores can use for scaling.

PeterHjortLauritzen replied:
Yes and higher value without Topo:

      initial_global_ave_dry_ps = 98288.0_r8
      if (.not. associated(fh_topo)) then
         initial_global_ave_dry_ps = 101325._r8 - 245._r8

MiCurry commented on src/dynamics/mpas/dyn_comp.F90:
```diff
@@ -1028,6 +1045,69 @@ subroutine read_inidat(dyn_in)
+                            + 0.5_r8*(zint(k+1,i)-zint(k  ,i))*gravity*rho(k  ,i)*(1.0_r8+tracers(ixqv,k  ,i))
+         end do
+
+      ! (7) theta_m from the state equation, compute rho_zz and theta while we are here                                                           
```
I believe this comment should be: "Compute theta_m from the state equation", correct @skamaroc?

MiCurry commented:
> Great. Feel free to use the FV3 dyn_comp.F90 as a template for types of logging messages. They will help us validate the additional functionality. Sounds like Peter is OK with adding these pressures in physconst.F90, maybe ps_dry_topo, ps_dry_notopo ! CESM reference dry mass pressure with/without topography. I can change them during the PR to cam_development if needed.

Sounds good. I moved this code to a new subroutine today, removed the dead code and added the log messages in your branch, I'll double check and see if the log messages are the same with the other dycores and then I'll add an additional commit which adds to `physconst.F90`.

MiCurry committed:
Move dry air mass scaling to new function and clean-up

This commit moves the dry air mass scaling changes inplmented in the proceeding
commits to a new subroutine, set_dry_mass. This is to be congruent with other
dycores (SE, FV3), which also have a similarly named function.

This commit also removes unsued code and adds in additional logging messages
similar to other dycores.

MiCurry committed:
Add CESM reference dry mass pressure to physconst

This commit adds two dry mass reference pressure to physconst. One for
topography and one for without topography. Dycores can use this to scale dry
air mass, but as of this commit no-dycores use it.

MiCurry committed:
Update MPAS dycore set_dry_mass to physconst reference dry mass pressure

This commit updates the set_dry_mass subroutine in the MPAS dycore to use the
CESM reference pressure added in the previous commit. Depending on if there is
topography, a different reference pressure is used and passed into
set_dry_mass.

The reference pressure is then used calcuated a scaling ratio which is used to
rescale the dry air mass.

MiCurry committed:
Add more log messages to set_dry_mass and compute scaled surface pressure

This commit adds additional log messages to the set_dry_mass routine and
rewrites previous log messages to be similar to other dycores.

In one of the log messages above, it logs the average dry surface
pressure after it is scaled, the computation of which is added in this commit.
This is similar to the output of other dycores.

MiCurry commented:
@jtruesdal @PeterHjortLauritzen I have just pushed four additional commits that: 1.) Moves the dry mass scaling to a new function, set_dry_mass, 2.) Add the physical constants mentioned above to physconst.F90 and uses them in the MPAS dycore and 3.) adds in additional log messages, which includes calculating the scaled dry air mass (taken from John's Fork).

Please feel free to give it another review. Are there any other log messages you think should be added?

MiCurry commented:
@jtruesdal Is there a namelist group you recommend adding `scale_dry_air_mass` so that it can be easily accessible by all dycores? Perhaps, `ref_pres_nl`?

jtruesdal commented:
We have two that I think would work 1) phys_ctl_nl   2) air_composition_nl which specifies wet/dry air compositions. 
 @PeterHjortLauritzen what would you like?

jtruesdal commented:
Actually ref_pres_nl sounds like the natural place for this.  I'm behind the curve, I didn't even know that existed.

PeterHjortLauritzen commented:
I am not sure ref_pres_nl is the right place when looking at the variables in that namelist. It almost seems like we need an initialization namelist?

MiCurry commented:
@PeterHjortLauritzen @jtruesdal There is the `cam_initfiles_nl`, which has `pertlim`, which perturbs the temperature up by a given amount. Its also defaults to 0.0.

PeterHjortLauritzen commented:
Great suggestion. What do you think @jtruesdal?

jtruesdal commented:
Thats it!

MiCurry committed:
Add dry air scaling nl option to cam_initfiles_nl

This commit removes the logical, mpas_scale_dry_air_mass namelist option and
replaces it a real-valued, scale_dry_air_mass nl option. Instead of having the
scale_dry_air_mass option in nhyd_model namelist group, it has been moved to
cam_initfiles_nl so that it can later be used by other dycores.

Currently, the namelist is only used by the MPAS dycore.

This commit also updates the calling logic in the mpas dycore so that
set_dry_mass is not called when running with simple physics and if the
scale_dry_air_mass value is 0.0.

MiCurry commented:
~Okay, @jtruesdal and @PeterHjortLauritzen, I just pushed an additional commit which adds the nl option, scale_dry_air_mass, and changes the logic of the calling code for the `set_dry_mass`.~

~@PeterHjortLauritzen, I updated the code from your pseudocode so that there is only one call to`set_dry_mass` (rather than two calls in separate locations).~

~Does the namelist description for `scale_dry_air_mass` look appropriate to you?~

Actually looks like this doesn't run. I'll update y'all with changes with it tomorrow. @jtruesdal is it failing to run because I'm reading the namelist wrong?

jtruesdal commented:
@MiCurry Looks like you forgot to add scale_dry_air_mass to namelist /cam_initfiles_nl/ in cam_initfiles.F90.  It will need to be defined as well as mpi_bcast like the rest.

jt

MiCurry commented:
Okay, sorry for the delay @jtruesdal and @PeterHjortLauritzen, I just pushed an update to this branch which, instead of reading the cam_initfiles_nl in the MPAS dycore, adds `scale_dry_air_mass` namelist variable as a public variable in `cam_initfiles.F90`, which each dycore can access on its on with `use`.

I also added units to `ps_dry_topo/notopo` in `physconts.F90`, I did so with a `git amend`, so if you have this branch checked out, you'll need to reset your branch to match my remote, which you can do by:

```
git fetch curry #Fetch my remote
git checkout cam-mpas/mass_scaling # If its not already checked out
git reset --hard curry/cam-mpas/mass_scaling # Update your branch to match my remote
```

I tested this branch successfully with different values of scale_dry_air_mass (i.e. < 0, 0.0, ~98000.0)

Let me know if you have any questions or problems.

jtruesdal commented:
Thanks Miles I'm testing it now.

jt
